### PR TITLE
!!! TASK: Rename `NodeTypeConstraints` to `NodeTypeCriteria`

### LINKED ISSUE #4522: !!! Move and possibly rename `NodeTypeConstraints`
As discussed in slack https://neos-project.slack.com/archives/C3MCBK6S2/p1694797399762309 the naming of `NodeTypeConstraints` when used as `"instanceof Foo.Bar"` filter has naming wise some overlap to actual NodeType Constraints https://docs.neos.io/guide/manual/content-repository/nodetype/node-constraints

```yaml
'Neos.Demo:Document.Chapter':
  constraints:
    nodeTypes:
      'Neos.Neos:Document': false
      'Neos.Demo:Document.Chapter': true
```

This might not be ideal and we should maybe discuss alternative namings, while we can ^^
Early suggestions include: `NodeTypeInstanceOf` `NodeTypeAllowList` (no they are not ideal! ^^)

A second more clear task is to move the `NodeTypeConstraints` (and the internal `NodeTypeConstraintsWithSubNodeTypes`) from the `Neos\ContentRepository\Core\Projection\ContentGraph`  namespace two level deeper (`Filter\NodeTypeConstraints`) to the other filters: `Neos\ContentRepository\Core\Projection\ContentGraph\Filter\NodeTypeConstraints`

### PR DESCRIPTION
**Pull request by bwaidelich**
Fixes: #4522

### CODE REVIEW

**Comment by bwaidelich** [comment]
*Notes:* Needs https://github.com/neos/neos-ui/pull/3625 to be merged as well

**Review by mhsdesign — APPROVED** [review]
tested locally with the ui: i ran the neos-ui e2e tests locally and it seems to work fine ;)

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php [hunk 1]
```diff
@@ -217,7 +217,7 @@ protected function requireNodeTypeToAllowNodesOfTypeInReference(
             throw ReferenceCannotBeSet::becauseTheNodeTypeDoesNotDeclareIt($referenceName, $nodeTypeName);
         }
         if (isset($propertyDeclaration['constraints']['nodeTypes'])) {
-            $nodeTypeConstraints = NodeTypeConstraintsWithSubNodeTypes::createFromNodeTypeDeclaration(
+            $nodeTypeConstraints = ExpandedNodeTypeCriteria::createFromNodeTypeDeclaration(
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php]: The `ExpandedNodeTypeCriteria` is internal and should not be used outside of the corresponding sub package

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php]: Solved in https://github.com/neos/neos-development-collection/pull/4679 ;)

**Review by ahaeslich — APPROVED** [review]

**Pull request merged**